File: python/wolfram_app.py
import tkinter as tk
from tkinter import messagebox, font
import wolframalpha
import os
from dotenv import load_dotenv

load_dotenv()

def initialize():
    app_id = os.getenv("WOLFRAM_FULL_RESPONSE")
    if not app_id:
        raise ValueError("WOLFRAM_FULL_RESPONSE is not set")

    return wolframalpha.Client(app_id)

def query(client: wolframalpha.Client):
    query_text = input_field.get()
    if not query_text.strip():
        messagebox.showinfo("Input Error", "Please enter a query.")
        return
    try:
        res = client.query(query_text)
        answer = next(res.results).text
        result_box.config(text=answer)
    except Exception:
        result_box.config(text="Error: Could not retrieve answer.")

def on_enter(event):
    query(client)

if __name__ == "__main__":
    client = initialize()
    root = tk.Tk()
    root.title("Wolfram App")
    root.geometry("600x300")
    root.configure(bg="#f0f0f0")

    # Customize fonts
    header_font = font.Font(family="Helvetica", size=16, weight="bold")
    label_font = font.Font(family="Arial", size=12)

    # Header Label
    header_label = tk.Label(root, text="Wolfram Query App", font=header_font, bg="#f0f0f0", fg="#333")
    header_label.grid(row=0, column=0, columnspan=2, pady=(20, 10), padx=10)

    # Input Label and Entry Field
    input_label = tk.Label(root, text="Enter your query:", font=label_font, bg="#f0f0f0")
    input_label.grid(row=1, column=0, pady=10, padx=10, sticky="e")

    input_field = tk.Entry(root, width=45, font=label_font)
    input_field.grid(row=1, column=1, pady=10, padx=10)
    input_field.bind("<Return>", on_enter)  # Bind Enter key to submit query

    # Query Button
    query_button = tk.Button(root, text="Query", command=lambda: query(client), bg="#4CAF50", fg="white", font=label_font)
    query_button.grid(row=2, column=0, columnspan=2, pady=10)

    # Result Box - Light Gray Background
    result_box = tk.Label(root, text="", font=label_font, bg="#e0e0e0", width=60, height=4, wraplength=550, justify="left", anchor="nw", padx=10, pady=10)
    result_box.grid(row=3, column=0, columnspan=2, pady=(10, 20), padx=10)

    
    root.mainloop()

File: go/wolframApp.go
package main

import (
	"encoding/xml"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"io"
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/widget"
	"github.com/joho/godotenv"
)

// Struct to parse Wolfram Alpha API response
type WolframResponse struct {
    XMLName xml.Name `xml:"queryresult"`
    Success bool     `xml:"success,attr"`
    Pods    []Pod    `xml:"pod"`
}

type Pod struct {
    Title  string `xml:"title,attr"`
    SubPods []SubPod `xml:"subpod"`
}

type SubPod struct {
    Text string `xml:"plaintext"`
}

// Function to initialize Wolfram Alpha API client
func initializeClient() string {
	// Load API key from .env file
	err := godotenv.Load()
	if err != nil {
		fmt.Println("Error loading .env file")
	}
	apiKey := os.Getenv("WOLFRAM_FULL_RESPONSE")
	if apiKey == "" {
		panic("WOLFRAM_FULL_RESPONSE is not set in .env file")
	}
	return apiKey
}

// Function to perform a query
func queryWolfram(query string) (string, error) {
    appID := initializeClient()
    encodedQuery := url.QueryEscape(query)
    url := fmt.Sprintf("http://api.wolframalpha.com/v2/query?input=%s&appid=%s", encodedQuery, appID)

    // Make the HTTP request
    resp, err := http.Get(url)
    if err != nil {
        return "", fmt.Errorf("error making request: %w", err)
    }
    defer resp.Body.Close()

    // Log status and body for debugging
    body, err := io.ReadAll(resp.Body)
    if err != nil {
        return "", fmt.Errorf("error reading response body: %w", err)
    }

    // Parse XML response
    var result WolframResponse
    if err := xml.Unmarshal(body, &result); err != nil {
        return "", fmt.Errorf("error parsing XML: %w", err)
    }

    // Check if the query was successful and look for the "Result" pod first
    if result.Success {
        // Try to find the "Result" pod with the calculated answer
        for _, pod := range result.Pods {
            if pod.Title == "Result" && len(pod.SubPods) > 0 {
                return pod.SubPods[0].Text, nil
            }
        }
        // If "Result" pod is not available, fallback to any available pod with plaintext
        for _, pod := range result.Pods {
            if len(pod.SubPods) > 0 && pod.SubPods[0].Text != "" {
                return pod.SubPods[0].Text, nil
            }
        }
    }
    return "No result found.", nil
}

func main() {
	// Create the main application window
	myApp := app.New()
	myWindow := myApp.NewWindow("Wolfram Alpha")
	myWindow.Resize(fyne.NewSize(600, 300))

	// Create input field and result label
	inputField := widget.NewEntry()
	inputField.SetPlaceHolder("Enter your query here...")

	resultBox := widget.NewLabel("")
	resultBox.Wrapping = fyne.TextWrapWord

	// Function to handle button click
	queryButton := widget.NewButton("Query", func() {
		query := inputField.Text
		if query == "" {
			resultBox.SetText("Please enter a query.")
			return
		}
		answer, err := queryWolfram(query)
		if err != nil {
			resultBox.SetText("Error: Could not retrieve answer.")
		} else {
			resultBox.SetText(fmt.Sprint(answer))
		}
	})

	// Set up the layout
	content := container.NewVBox(
		widget.NewLabelWithStyle("Wolfram Query App", fyne.TextAlignCenter, fyne.TextStyle{Bold: true}),
		inputField,
		queryButton,
		widget.NewSeparator(),
		resultBox,
	)

	// Set up the Enter key press to trigger the query
	inputField.OnSubmitted = func(text string) {
		queryButton.OnTapped()
	}

	myWindow.SetContent(content)
	myWindow.ShowAndRun()
}
